src/generic_factory.h

```diff
@@ -223,25 +296,63 @@ class generic_factory
         /**
          * Returns the object with the given id.
          * The input id should be valid, otherwise a debug message is issued.
+         * This function can be used to implement @ref int_id::obj().
+         * Note: If the id was valid, the returned object can be modified (after
+         * casting the const away).
+         */
+        const T &obj( const int_id<T> &id ) const {
+            if( !is_valid( id ) ) {
+                debugmsg( "invalid %s id \"%d\"", type_name.c_str(), id );
+                return dummy_obj();
+            }
+            return list[id];
+        }
+        /**
+         * Returns the object with the given id.
+         * The input id should be valid, otherwise a debug message is issued.
          * This function can be used to implement @ref string_id::obj().
          * Note: If the id was valid, the returned object can be modified (after
          * casting the const away).
          */
-        const T &obj( const string_id<T> &sid ) const {
-            const auto iter = data.find( sid );
-            if( iter == data.end() ) {
-                debugmsg( "invalid %s id \"%s\"", type_name.c_str(), sid.c_str() );
-                static const T dummy{};
-                return dummy;
+        const T &obj( const string_id<T> &id ) const {
+            static int_id<T> i_id;
+            if( !find_id( id, i_id ) ) {
+                debugmsg( "invalid %s id \"%s\"", type_name.c_str(), id.c_str() );
+                return dummy_obj();
```

Comment:
This needs to be a static variable. Currently this returns a reference to a temporary object.